### PR TITLE
`CalcJob`: always call `Scheduler.parse_output`

### LINKED ISSUE #4840: Parse stdout and stderr from schedulers even if detailed job info is not available
When we implemented the `Scheduler.parse_output` functionality to let scheduler plugins parse the scheduler output in order to detect potential problems, we decided to not rely exclusively on the text in the stdout and stderr streams, since parsing information reliably from these unstructured text sources is error prone and unreliable. For SLURM, OOM and OOW problems are detected exclusively from the detailed job info, which is reported by the `sacct` command. However, this is not a standard feature for the SLURM scheduler and not all installations have it. This causes OOW and OOM errors to get passed unnoticed for many installations. The question is now if we should relax this strict condition of requiring the `detailed_job_info` in the `Scheduler.parse_output` call and allow to also parse from the text.

@adegomme suggested the following regexs for a number of schedulers in [this issue](https://github.com/BigDFT-group/aiida-bigdft-plugin/issues/3):

Slurm : "CANCELLED AT xx DUE TO TIME LIMIT" https://github.com/SchedMD/slurm/blob/master/src/slurmd/slurmstepd/req.c#L747
PBS : "PBS: job killed: walltime xx exceeded limit " https://github.com/openpbs/openpbs/blob/master/src/resmom/mom_main.c#L6040
SGE/UGE might be "exceeded hard wallclock time" https://github.com/gridengine/gridengine/blob/master/source/daemons/execd/msg_execd.h#L220
LSF : "TERM_RUNLIMIT: job killed after reaching LSF run time limit" https://www.ibm.com/support/pages/termrunlimit-job-killed-after-reaching-lsf-run-time-limit-exited-exit-code-140
Torque : "walltime xx exceeded limit xx" https://github.com/adaptivecomputing/torque/blob/master/src/resmom/mom_main.c#L3864

Implementing these should hopefully just add a lot of chances of detecting OOW even though some might still be missed if the format is different for other versions of the schedulers. At least, hopefully it should not report false positives.

The one technical question is that we would have to change the interface. Currently we have:
```python
def parse_output(self, detailed_job_info, stdout, stderr):
```
so all arguments are required, but we would have to make them all optional. I am not sure how many plugins outside of `aiida-core` already rely on this, but we should be careful with breaking this. Maybe it would be possible to pass empty defaults if any of them could not be retrieved, i.e. an empty dict for `detailed_job_info` and an empty string for `stdout` and `stderr`?

Pinging  @mbercx @giovannipizzi

### PR DESCRIPTION
**Pull request by sphuber**
Fixes #4840

When the functionality was added for a `Scheduler` plugin to parse the
output that was written to `stdout` and `stderr` or retrieved from a
specialized status command, it was decided to only call `parse_output`
if all three information streams were successfully retrieved. The
reasoning was especially that the `detailed_info` should be required as
that would return well-structured data and would allow to reliably
determine what had happened, whereas parsing free text from stdout and
stderr would be error prone.

Although probably a safe choice, the direct result was that setups where
the scheduler didn't have the necessary implementation to return the
`detailed_info`, no scheduler output parsing would be available. This
would lead to many OOM and OOW problems to go unnoticed and the engine
retrying to submit without any error handling.

Here we remove the requirement that all three information streams from
the parser, `detailed_info`, `stderr` and `stdout` should be known, but
that `parse_output` will always be called, with a default of `None` for
each. The advantage is that this will allow scheduler plugins to
implement parsing from `stderr`, on top of `detailed_info` wherever
applicable, to increase the chances of catching basic problems.

The major downside is that this is a backwards incompatible change for
scheduler plugins that rely on the assumption that the arguments always
have values that are not `None`.

### CODE REVIEW

File: aiida/engine/processes/calcjobs/calcjob.py
Comment (ltalirz) [inline aiida/engine/processes/calcjobs/calcjob.py:531]: I think we could replace this with

```python
if all(entry is None for entry in [detailed_job_info, scheduler_stderr, scheduler_stdout]):
```

File: aiida/schedulers/plugins/slurm.py
Comment (ltalirz) [inline aiida/schedulers/plugins/slurm.py:719]: I guess the check on the `detailed_job_info` is scheduler specific (?).

Checks on `stdout` or `stderr` (i.e. `str` or `None`) could be done at a general level, though, if we like.

File: aiida/schedulers/plugins/slurm.py
Comment (ltalirz) [inline aiida/schedulers/plugins/slurm.py:710]: Just mentioning that we don't necessarily need to provide default values here (but I guess this makes it explicit that these parameters can be `None`)

File: aiida/engine/processes/calcjobs/calcjob.py
Comment (ltalirz) [reply aiida/engine/processes/calcjobs/calcjob.py:531]: Furthermore, while I understand that `detailed_job_info` can be `None`, I wonder whether it is ever legitimate for `stdout` or `stderr` to be `None`?

`node.get_option()` will return `None`, if the option is missing, but is that not an indication that something is wrong?
Perhaps a warning should be logged in that case (not necessary for this PR, though).

**Review by ltalirz — CHANGES_REQUESTED** [review]
Fantastic, thanks a lot @sphuber !

Just some minor suggestions/questions

File: aiida/schedulers/plugins/slurm.py
Comment (sphuber) [reply aiida/schedulers/plugins/slurm.py:719]: We could, but I think it is better to keep this on the plugin level rather than collecting all possible regexes on the `Scheduler` base class.

File: aiida/engine/processes/calcjobs/calcjob.py
Comment (sphuber) [reply aiida/engine/processes/calcjobs/calcjob.py:531]: A warning is being logged if either is `None` on line 522 and 528 respectively. In the spirit of this PR, I think we should keep it at a warning and continue calling `parse_output` and leave it up to the plugins to try and extract as much information as possible.

Comment (sphuber) [reply aiida/engine/processes/calcjobs/calcjob.py:531]: But I think you make a good point about `node.get_option` returning `None` if the option was never set, because it is being passed straight to `retrieved.get_object_content` and this should result in a `TypeError` being thrown from `Repository.get_file` which doesn't accept `None`. This should also be caught.

**Review by ltalirz — APPROVED** [review]
thanks a lot!

File: aiida/schedulers/plugins/slurm.py
Comment (ltalirz) [inline aiida/schedulers/plugins/slurm.py:775]: Damn, just noticed this small error here.

Comment (chrisjsewell) [reply aiida/schedulers/plugins/slurm.py:775]: Eurgh, would it not have been good to add some tests in this PR, to catch things like this

Comment (ltalirz) [reply aiida/schedulers/plugins/slurm.py:775]: Yep; I think it's fair to do this in the follow-up PR that adds support for the remaining schedulers.
Did not want to add more work for @sphuber who jumped in quickly here

Comment (chrisjsewell) [reply aiida/schedulers/plugins/slurm.py:775]: Slow and steady wins the race 😅

Comment (ltalirz) [reply aiida/schedulers/plugins/slurm.py:775]: Nordic walking?

Comment (sphuber) [reply aiida/schedulers/plugins/slurm.py:775]: It also shouldn't have been squashed merged. I will correct both mistakes and force push

Comment (sphuber) [reply aiida/schedulers/plugins/slurm.py:775]: And I wasn't really thinking this was ready to be merged. Would have suggested to test this in the wild with SLURM to see if the regex patterns actually match. Had just taken them from the issue discussion, but not sure that these are actually correct.

Comment (chrisjsewell) [reply aiida/schedulers/plugins/slurm.py:775]: Yeh, I moaned at @sphuber about this before lol; for maintainers lets not merge each other's PRs, unless given "permission". At least I don't want people merging my PRs 😜

Comment (sphuber) [reply aiida/schedulers/plugins/slurm.py:775]: Anyway, all good now. Corrected the commits on `develop`.

Comment (ltalirz) [reply aiida/schedulers/plugins/slurm.py:775]: > for maintainers lets not merge each other's PRs, unless given "permission" 

Ok, will keep it in mind as a policy on this repository for the future.

I typically mark my PRs as "draft" if I don't want them to be merged yet, sorry for the confusion.

Comment (chrisjsewell) [reply aiida/schedulers/plugins/slurm.py:775]: It's not that they are not ready to be merged, it's that I want to make sure that it is merged correctly (squash/merge/rebase), and that the commit message is correct

**Pull request merged**
